src/upgrade.py

```diff
@@ -0,0 +1,219 @@
+        if self.charm._mysql.get_primary_label() != f"{self.charm.app.name}-0":
+            # set the primary to the first unit for switchover mitigation
+            self.charm._mysql.set_cluster_primary(
+                self.charm._get_unit_fqdn(f"{self.charm.app.name}/0")
+            )
+
+        self.charm.k8s_helpers.set_rolling_update_partition(partition=self.highest_ordinal)
+
+    def _on_resume_upgrade(self, event: ActionEvent) -> None:
+        """Handle resume upgrade action.
+
+        Continue the upgrade by setting the partition to the next unit.
+        """
+        if not self.charm.unit.is_leader():
+            event.fail("Action only valid on leader unit")
+            return
+        fail_message = "Nothing to resume, upgrade stack unset"
+        if self.upgrade_stack:
+            try:
+                next_partition = self.upgrade_stack[-1]
+                event.set_results({"message": f"Upgrade will resume on unit {next_partition}"})
+                self.charm.k8s_helpers.set_rolling_update_partition(partition=next_partition)
+                return
+            except IndexError:
+                fail_message = "Nothing to resume, empty upgrade stack"
+            except KubernetesClientError:
+                fail_message = "Cannot set rolling update partition"
+        event.fail(fail_message)
+
+    def _on_pebble_ready(self, event) -> None:
+        """Handle pebble ready event.
+
+        Confirm that unit is healthy and set unit completed.
+        """
+        if not self.peer_relation:
+            logger.debug("Peer relation not yet ready on unit. Deferring event")
+            event.defer()
+            return
+
+        if self.peer_relation.data[self.charm.unit].get("state") in ["ready", "upgrading"]:
+            try:
+                for attempt in Retrying(stop=stop_after_attempt(6), wait=wait_fixed(10)):
+                    with attempt:
+                        if self.charm._mysql.is_instance_in_cluster(self.charm.unit_label):
+                            logger.debug(
+                                "Upgraded unit is healthy. Set upgrade state to `completed`"
+                            )
+                            self.set_unit_completed()
+                            # call update status asap
+                            self.charm._on_update_status(None)
+                        else:
+                            logger.debug(
+                                "Instance not yet back in the cluster."
+                                f" Retry {attempt.retry_state.attempt_number}/6"
+                            )
+                            raise Exception
+            except RetryError:
+                logger.debug("Upgraded unit is not healthy")
+                self.set_unit_failed()
+                self.charm.unit.status = BlockedStatus(
+                    "upgrade failed. Check logs for rollback instruction"
+                )
+                self.log_rollback_instructions()
```

Comment:
I think this call is not needed as it is already being done by the lib on https://github.com/canonical/data-platform-libs/blob/2a5ac125be55c1e7583df0e1f373f2768699640a/lib/charms/data_platform_libs/v0/upgrade.py#L706.

Reply:
269aea6